### PR TITLE
Do not process workflow tokens in the runservice endpoint

### LINKED ISSUE #13849: [production][trigger_workflow#create] undefined method `payload' for nil:NilClass

    raise Token::Errors::MissingPayload, 'A payload ...
[See this exception on Errbit](http://errbit.opensuse.org/apps/5620ca2bdc71fa00b1000000/problems/633e39aa84bf46052743f21c)
# undefined method `payload&#39; for nil:NilClass

    raise Token::Errors::MissingPayload, &#39;A payload is required&#39; if @scm_webhook.payload.blank?
                                                                                ^^^^^^^^ #
### URL ###
[https://api.opensuse.org/trigger/runservice](https://api.opensuse.org/trigger/runservice)"
### Where ###
trigger#create

### Occured ###
Oct 06  2:12am

### Similar ###
5

## Params ##
~~~
{
  &quot;action&quot; =&gt; &quot;create&quot;,
  &quot;controller&quot; =&gt; &quot;trigger&quot;
}
~~~

## Session ##
~~~
{}
~~~

## Backtrace ##
~~~
25:  app/models/token/workflow.rb -> **call**
34:  app/controllers/trigger_controller.rb -> **block in create**
887:  app/models/user.rb -> **run_as**
31:  app/controllers/trigger_controller.rb -> **create**
6:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_controller/metal/basic_implicit_render.rb -> **send_action**
228:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/abstract_controller/base.rb -> **process_action**
30:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_controller/metal/rendering.rb -> **process_action**
42:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/abstract_controller/callbacks.rb -> **block in process_action**
106:  /GEM_ROOT/gems/activesupport-6.1.5.1/lib/active_support/callbacks.rb -> **run_callbacks**
41:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/abstract_controller/callbacks.rb -> **process_action**
22:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_controller/metal/rescue.rb -> **process_action**
34:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_controller/metal/instrumentation.rb -> **block in process_action**
203:  /GEM_ROOT/gems/activesupport-6.1.5.1/lib/active_support/notifications.rb -> **block in instrument**
24:  /GEM_ROOT/gems/activesupport-6.1.5.1/lib/active_support/notifications/instrumenter.rb -> **instrument**
203:  /GEM_ROOT/gems/activesupport-6.1.5.1/lib/active_support/notifications.rb -> **instrument**
33:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_controller/metal/instrumentation.rb -> **process_action**
249:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_controller/metal/params_wrapper.rb -> **process_action**
27:  /GEM_ROOT/gems/activerecord-6.1.5.1/lib/active_record/railties/controller_runtime.rb -> **process_action**
165:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/abstract_controller/base.rb -> **process**
39:  /GEM_ROOT/gems/actionview-6.1.5.1/lib/action_view/rendering.rb -> **process**
190:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_controller/metal.rb -> **dispatch**
254:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_controller/metal.rb -> **dispatch**
50:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/routing/route_set.rb -> **dispatch**
33:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/routing/route_set.rb -> **serve**
19:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/routing/mapper.rb -> **block in &lt;class:Constraints&gt;**
49:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/routing/mapper.rb -> **serve**
50:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/journey/router.rb -> **block in serve**
32:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/journey/router.rb -> **each**
32:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/journey/router.rb -> **serve**
842:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/routing/route_set.rb -> **call**
72:  /GEM_ROOT/gems/flipper-0.25.2/lib/flipper/middleware/memoizer.rb -> **memoized_call**
37:  /GEM_ROOT/gems/flipper-0.25.2/lib/flipper/middleware/memoizer.rb -> **call**
15:  config/initializers/prefer_xml_over_html.rb -> **call**
44:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/deflater.rb -> **call**
15:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/tempfile_reaper.rb -> **call**
27:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/etag.rb -> **call**
40:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/conditional_get.rb -> **call**
12:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/head.rb -> **call**
22:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/http/permissions_policy.rb -> **call**
19:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/http/content_security_policy.rb -> **call**
266:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/session/abstract/id.rb -> **context**
260:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/session/abstract/id.rb -> **call**
689:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/cookies.rb -> **call**
27:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/callbacks.rb -> **block in call**
98:  /GEM_ROOT/gems/activesupport-6.1.5.1/lib/active_support/callbacks.rb -> **run_callbacks**
26:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/callbacks.rb -> **call**
18:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/actionable_exceptions.rb -> **call**
32:  /GEM_ROOT/gems/airbrake-9.4.3/lib/airbrake/rack/middleware.rb -> **call!**
21:  /GEM_ROOT/gems/airbrake-9.4.3/lib/airbrake/rack/middleware.rb -> **call**
29:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/debug_exceptions.rb -> **call**
33:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/show_exceptions.rb -> **call**
18:  /GEM_ROOT/gems/lograge-0.12.0/lib/lograge/rails_ext/rack/logger.rb -> **call_app**
26:  /GEM_ROOT/gems/railties-6.1.5.1/lib/rails/rack/logger.rb -> **block in call**
99:  /GEM_ROOT/gems/activesupport-6.1.5.1/lib/active_support/tagged_logging.rb -> **block in tagged**
37:  /GEM_ROOT/gems/activesupport-6.1.5.1/lib/active_support/tagged_logging.rb -> **tagged**
99:  /GEM_ROOT/gems/activesupport-6.1.5.1/lib/active_support/tagged_logging.rb -> **tagged**
26:  /GEM_ROOT/gems/railties-6.1.5.1/lib/rails/rack/logger.rb -> **call**
81:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/remote_ip.rb -> **call**
81:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/remote_ip.rb -> **call**
19:  /GEM_ROOT/gems/request_store-1.5.1/lib/request_store/middleware.rb -> **call**
26:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/request_id.rb -> **call**
24:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/method_override.rb -> **call**
22:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/runtime.rb -> **call**
29:  /GEM_ROOT/gems/activesupport-6.1.5.1/lib/active_support/cache/strategy/local_cache_middleware.rb -> **call**
14:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/executor.rb -> **call**
18:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/lock.rb -> **call**
110:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/sendfile.rb -> **call**
142:  /GEM_ROOT/gems/actionpack-6.1.5.1/lib/action_dispatch/middleware/host_authorization.rb -> **call**
539:  /GEM_ROOT/gems/railties-6.1.5.1/lib/rails/engine.rb -> **call**
207:  /GEM_ROOT/gems/railties-6.1.5.1/lib/rails/railtie.rb -> **public_send**
207:  /GEM_ROOT/gems/railties-6.1.5.1/lib/rails/railtie.rb -> **method_missing**
74:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/urlmap.rb -> **block in call**
58:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/urlmap.rb -> **each**
58:  /GEM_ROOT/gems/rack-2.2.4/lib/rack/urlmap.rb -> **call**
107:  /usr/lib64/ruby/gems/3.1.0/gems/passenger-6.0.14/src/ruby_supportlib/phusion_passenger/rack/thread_handler_extension.rb -> **process_request**
157:  /usr/lib64/ruby/gems/3.1.0/gems/passenger-6.0.14/src/ruby_supportlib/phusion_passenger/request_handler/thread_handler.rb -> **accept_and_process_next_request**
110:  /usr/lib64/ruby/gems/3.1.0/gems/passenger-6.0.14/src/ruby_supportlib/phusion_passenger/request_handler/thread_handler.rb -> **main_loop**
419:  /usr/lib64/ruby/gems/3.1.0/gems/passenger-6.0.14/src/ruby_supportlib/phusion_passenger/request_handler.rb -> **block (3 levels) in start_threads**
113:  /usr/lib64/ruby/gems/3.1.0/gems/passenger-6.0.14/src/ruby_supportlib/phusion_passenger/utils.rb -> **block in create_thread_and_abort_on_exception**
~~~

## Environment ##
| Key        | Value      |
|------------|------------|
| program_name | Passenger RubyApp: /srv/www/obs/api (production) |
| HTTP_USER_AGENT | Python-urllib/3.6 |

### PR DESCRIPTION
**Pull request by danidoni**
It makes no sense to process Workflow Tokens in the `/trigger/runservice` endpoint, so we just 404 when a request like that comes in.

Fixes #13849 

## How to Test It

Having a Workflow Token with an id of 1, a `home:Admin` project and a `hello_world` package:

```
osc api -X POST "http://localhost:3000/trigger/runservice?id=1&project=home:Admin&package=hello_world" -a X-Obs-Signature <secret>
```
or

Using the `/apidocs/index.html#/Trigger/post_trigger__operation_`

### CODE REVIEW

File: src/api/app/controllers/trigger_controller.rb [hunk 1]
```diff
@@ -42,6 +43,11 @@ def create
 
   private
 
+  def check_token_class
+    # It make no sense to process Token::Workflows on a runservice request
+    raise ActiveRecord::RecordNotFound, "Token not found" if @token.class == Token::Workflow
```
Comment (dmarcoux) [inline src/api/app/controllers/trigger_controller.rb:48]: I think this error message is somewhat confusing. Pointing users to the right endpoint would be helpful in this case.

Comment (eduardoj) [reply src/api/app/controllers/trigger_controller.rb:48]: This check seems very similar to the one performed in this line, belonging to a yet not merged pull request: https://github.com/openSUSE/open-build-service/pull/15119/files#diff-77ed5d907b983934be02351099c20880d01bca6aab8366863839fc0721b3bb5fR61

Comment (danidoni) [reply src/api/app/controllers/trigger_controller.rb:48]: Well, we don't know if the user chose the wrong endpoint or the wrong token id, so pointing to another endpoint will make things more confusing if the token type is not the right one...

Comment (dmarcoux) [reply src/api/app/controllers/trigger_controller.rb:48]: > we don't know if the user chose the wrong endpoint or the wrong token id

True, I didn't think about this at first.

File: src/api/app/controllers/trigger_workflow_controller.rb [hunk 2]
```diff
@@ -3,7 +3,7 @@ class TriggerWorkflowController < TriggerController
   include ScmWebhookPayloadDataExtractor
 
   # We don't need to validate that the body of the request is XML. We receive JSON
-  skip_before_action :validate_xml_request, :set_project_name, :set_package_name, :set_project, :set_package, :set_object_to_authorize, :set_multibuild_flavor
+  skip_before_action :validate_xml_request, :set_project_name, :set_package_name, :set_project, :set_package, :set_object_to_authorize, :set_multibuild_flavor, :check_token_class
```
Comment (danidoni) [inline src/api/app/controllers/trigger_workflow_controller.rb:6]: I would like to fix this whole line [here](https://trello.com/c/s3rXELVj/316-stop-using-inheritance-on-workflowtriggercontroller)

**Review by dmarcoux — APPROVED** [review]

**Pull request merged**